# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/cairo/cairo/src/cairo-script-surface.c
# Commit: 3cdfb39376b7
# Full Hash: 3cdfb39376b736597fe658a322389117686b7e82
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2024-05-07 09:13:07
# Regressor Bug: 1892913
# File Overlap Count: 3
# Description:
#   Bug 1892913 - patch 1 - Update cairo source to 1.18.0 r=gfx-reviewers,lsalzman
#   
#   Wholesale replacement of files in gfx/cairo/cairo with contents of the 1.18.0 release tarball,
#   omitting subdirectories that are not part of the core library build.
#   
# ==============================================================================

diff -r 9daa351f875b -r 3cdfb39376b7 gfx/cairo/cairo/src/cairo-script-surface.c
--- a/gfx/cairo/cairo/src/cairo-script-surface.c	Mon May 06 19:22:35 2024 +0000
+++ b/gfx/cairo/cairo/src/cairo-script-surface.c	Mon May 06 19:25:12 2024 +0000
@@ -371,6 +371,21 @@
 }
 
 static const char *
+_dither_to_string (cairo_dither_t dither)
+{
+    static const char *names[] = {
+	"DITHER_DEFAULT",	/* CAIRO_FILTER_FAST */
+	"DITHER_NONE",		/* CAIRO_FILTER_GOOD */
+	"DITHER_FAST",		/* CAIRO_FILTER_BEST */
+	"DITHER_GOOD",		/* CAIRO_FILTER_NEAREST */
+	"DITHER_BEST",		/* CAIRO_FILTER_BILINEAR */
+    };
+    assert (dither < ARRAY_LENGTH (names));
+    return names[dither];
+}
+
+
+static const char *
 _fill_rule_to_string (cairo_fill_rule_t rule)
 {
     static const char *names[] = {
@@ -709,6 +724,24 @@
 }
 
 static cairo_status_t
+_emit_hairline (cairo_script_surface_t *surface, cairo_bool_t set_hairline)
+{
+    assert (target_is_active (surface));
+
+    if (surface->cr.current_style.is_hairline == set_hairline)
+    {
+	return CAIRO_STATUS_SUCCESS;
+    }
+
+    surface->cr.current_style.is_hairline = set_hairline;
+
+    _cairo_output_stream_printf (to_context (surface)->stream, 
+					"%d set-hairline\n",
+					set_hairline);
+    return CAIRO_STATUS_SUCCESS;
+}
+
+static cairo_status_t
 _emit_line_cap (cairo_script_surface_t *surface,
 		cairo_line_cap_t line_cap)
 {
@@ -858,6 +891,10 @@
     if (unlikely (status))
 	return status;
 
+    status = _emit_hairline (surface, style->is_hairline);
+    if (unlikely (status))
+	return status;
+
     status = _emit_dash (surface,
 			 style->dash, style->num_dashes, style->dash_offset,
 			 force);
@@ -1709,6 +1746,17 @@
 				     " //%s set-filter\n ",
 				     _filter_to_string (pattern->filter));
     }
+    /* XXX need to discriminate the user explicitly setting the default */
+    if (pattern->dither != CAIRO_DITHER_DEFAULT) {
+	if (need_newline) {
+	    _cairo_output_stream_puts (ctx->stream, "\n ");
+	    need_newline = FALSE;
+	}
+
+	_cairo_output_stream_printf (ctx->stream,
+				     " //%s set-dither\n ",
+				     _dither_to_string (pattern->dither));
+    }
     if (! is_default_extend ){
 	if (need_newline) {
 	    _cairo_output_stream_puts (ctx->stream, "\n ");
@@ -1860,7 +1908,7 @@
 
     if (! _cairo_path_fixed_iter_at_end (&iter)) {
 	_cairo_boxes_fini (&boxes);
-	return CAIRO_STATUS_INVALID_PATH_DATA;
+	return CAIRO_INT_STATUS_UNSUPPORTED;
     }
 
     for (chunk = &boxes.chunks; chunk; chunk = chunk->next) {
@@ -2091,6 +2139,16 @@
 }
 
 static void
+_device_finish (void *abstract_device)
+{
+    cairo_script_context_t *ctx = abstract_device;
+
+    cairo_status_t status = _cairo_output_stream_close (ctx->stream);
+    status = _cairo_device_set_error (&ctx->base, status);
+    (void) status;
+}
+
+static void
 _device_destroy (void *abstract_device)
 {
     cairo_script_context_t *ctx = abstract_device;
@@ -2172,6 +2230,7 @@
 
     _cairo_pattern_fini (&surface->cr.current_source.base);
     _cairo_path_fixed_fini (&surface->cr.current_path);
+    _cairo_font_options_fini (&surface->cr.current_font_options);
     _cairo_surface_clipper_reset (&surface->clipper);
 
     status = cairo_device_acquire (&ctx->base);
@@ -2207,8 +2266,7 @@
 		    cairo_list_del (&surface->operand.link);
 		} else {
 		    link->operand.type = DEFERRED;
-		    cairo_list_swap (&link->operand.link,
-				     &surface->operand.link);
+		    cairo_list_move_list (&surface->operand.link, &link->operand.link);
 		    cairo_list_add (&link->link, &ctx->deferred);
 		}
 	    }
@@ -2477,7 +2535,7 @@
 
     if (_cairo_surface_wrapper_is_active (&surface->wrapper)) {
 	return _cairo_surface_wrapper_paint (&surface->wrapper,
-					     op, source, clip);
+					     op, source, 0, clip);
     }
 
     return CAIRO_STATUS_SUCCESS;
@@ -2534,7 +2592,7 @@
 
     if (_cairo_surface_wrapper_is_active (&surface->wrapper)) {
 	return _cairo_surface_wrapper_mask (&surface->wrapper,
-					    op, source, mask, clip);
+					    op, source, 0, mask, 0, clip);
     }
 
     return CAIRO_STATUS_SUCCESS;
@@ -2621,7 +2679,7 @@
 
     if (_cairo_surface_wrapper_is_active (&surface->wrapper)) {
 	return _cairo_surface_wrapper_stroke (&surface->wrapper,
-					      op, source, path,
+					      op, source, 0, path,
 					      style,
 					      ctm, ctm_inverse,
 					      tolerance, antialias,
@@ -2702,7 +2760,7 @@
 
     if (_cairo_surface_wrapper_is_active (&surface->wrapper)) {
 	return _cairo_surface_wrapper_fill (&surface->wrapper,
-					    op, source, path,
+					    op, source, 0, path,
 					    fill_rule,
 					    tolerance,
 					    antialias,
@@ -3012,6 +3070,7 @@
 	if (unlikely (status))
 	    return status;
 
+	_cairo_font_options_init_default (&options);
 	cairo_scaled_font_get_font_options (scaled_font, &options);
 	status = _emit_font_options (surface, &options);
 	if (unlikely (status))
@@ -3047,7 +3106,7 @@
 
     index = ++font_private->subset_glyph_index;
     scaled_glyph->dev_private_key = ctx;
-    scaled_glyph->dev_private = (void *) index;
+    scaled_glyph->dev_private = (void *)(uintptr_t)index;
 
     _cairo_output_stream_printf (ctx->stream,
 				 "%lu <<\n"
@@ -3095,7 +3154,7 @@
 
     index = ++font_private->subset_glyph_index;
     scaled_glyph->dev_private_key = ctx;
-    scaled_glyph->dev_private = (void *) index;
+    scaled_glyph->dev_private = (void *)(uintptr_t)index;
 
     _cairo_output_stream_printf (ctx->stream,
 				 "%lu <<\n"
@@ -3172,6 +3231,7 @@
 	status = _cairo_scaled_glyph_lookup (scaled_font,
 					     glyphs[n].index,
 					     CAIRO_SCALED_GLYPH_INFO_METRICS,
+                                             NULL, /* foreground color */
 					     &scaled_glyph);
 	if (unlikely (status))
 	    break;
@@ -3182,6 +3242,7 @@
 	status = _cairo_scaled_glyph_lookup (scaled_font,
 					     glyphs[n].index,
 					     CAIRO_SCALED_GLYPH_INFO_RECORDING_SURFACE,
+                                             NULL, /* foreground color */
 					     &scaled_glyph);
 	if (_cairo_status_is_error (status))
 	    break;
@@ -3207,6 +3268,7 @@
 	status = _cairo_scaled_glyph_lookup (scaled_font,
 					     glyphs[n].index,
 					     CAIRO_SCALED_GLYPH_INFO_SURFACE,
+                                             NULL, /* foreground color */
 					     &scaled_glyph);
 	if (_cairo_status_is_error (status))
 	    break;
@@ -3290,7 +3352,7 @@
 	    _cairo_output_stream_printf (ctx->stream, "\\%c", c);
 	    break;
 	default:
-	    if (isprint (c) || isspace (c)) {
+	    if (_cairo_isprint(c)) {
 		_cairo_output_stream_printf (ctx->stream, "%c", c);
 	    } else {
 		char buf[4] = { '\\' };
@@ -3389,13 +3451,14 @@
 	    status = _cairo_scaled_glyph_lookup (scaled_font,
 						 glyphs[n].index,
 						 CAIRO_SCALED_GLYPH_INFO_METRICS,
+						 NULL, /* foreground color */
 						 &scaled_glyph);
 	    if (unlikely (status)) {
 		_cairo_scaled_font_thaw_cache (scaled_font);
 		goto BAIL;
 	    }
 
-	    if ((long unsigned) scaled_glyph->dev_private > 256)
+	    if ((uintptr_t)scaled_glyph->dev_private > 256)
 		break;
 	}
     }
@@ -3412,6 +3475,7 @@
 	status = _cairo_scaled_glyph_lookup (scaled_font,
 					     glyphs[n].index,
 					     CAIRO_SCALED_GLYPH_INFO_METRICS,
+                                             NULL, /* foreground color */
 					     &scaled_glyph);
 	if (unlikely (status)) {
 	    _cairo_scaled_font_thaw_cache (scaled_font);
@@ -3466,7 +3530,7 @@
 	    if (font_private->has_sfnt)
 		c = glyphs[n].index;
 	    else
-		c = (uint8_t) (long unsigned) scaled_glyph->dev_private;
+		c = (uint8_t) (uintptr_t) scaled_glyph->dev_private;
 
 	    _cairo_output_stream_write (base85_stream, &c, 1);
 	} else {
@@ -3475,7 +3539,7 @@
 					     glyphs[n].index);
 	    else
 		_cairo_output_stream_printf (ctx->stream, " %lu",
-					     (long unsigned) scaled_glyph->dev_private);
+					     (long unsigned) (uintptr_t)scaled_glyph->dev_private);
 	}
 
         dx = scaled_glyph->metrics.x_advance;
@@ -3548,7 +3612,7 @@
 
     if (_cairo_surface_wrapper_is_active (&surface->wrapper)){
 	return _cairo_surface_wrapper_show_text_glyphs (&surface->wrapper,
-							op, source,
+							op, source, 0,
 							utf8, utf8_len,
 							glyphs, num_glyphs,
 							clusters, num_clusters,
@@ -3704,7 +3768,7 @@
     NULL, NULL, /* lock, unlock */
 
     _device_flush,  /* flush */
-    NULL,  /* finish */
+    _device_finish,  /* finish */
     _device_destroy
 };
 
@@ -3926,7 +3990,6 @@
 						   content, extents,
 						   NULL)->base;
 }
-slim_hidden_def (cairo_script_surface_create);
 
 /**
  * cairo_script_surface_create_for_target: